### ISSUE TITLE
Bandstructure handling

### ISSUE THREAD

**Issue by shyuep** [materialsproject/pymatgen#638]
I am thinking that the Bandstructure capabilities need a bit of an overhaul. Recently, I have tried to plot a few bandstructures from a few vasp runs (not from MP), but found that I had to play around with several parameters until I finally got it to work. For example,

1. Vasprun.get_band_structure returns two types of Bandstructures, only one (SymmLine) of which BSPlotter would take. To get a SymmLine, I had to specify both kpoints file and line_mode=True. For some reason, supplying the Kpoints file itself is insufficient to get it to detect it is a symm line version, even though all the labels have been generated from MPNonSCFSet. 
2. Why can't the get_band_structure method autodetect whether kpoints are along symmetry lines, whether the labels are specified or not? The labels should be generated from symmetry analysis right? 
3. In the pymatgen example notebooks, there are no examples of how you would go about running a BS calculation and plotting that band structure. There is only an example of getting a pre-computed BS from MP and plotting this. I think the former is very important for users.
4. BS reading still gives a "Structure is not standard primitive", even though the structure was generated using get_primitive_standard. My conclusion is that the symprec default used in the symmetry.Bandstructure is 0.01, which is different from the defaults in other cases.

@hautierg @computron Can you guys comment and can some improvements be made?

**Comment by computron** [comment]
Try BSDOSPlotter, that's the one I wrote and IMO much prettier than BSPlotter

**Comment by shyuep** [comment]
Erm, that does not solve the SymmLine vs non-SymmLine issue?

**Comment by computron** [comment]
It only solves the issue of needing to play around with a lot of parameters to get the plotting to work. You should just be able to feed in a DOS and BSSymmLine object. If you are feeding in a non symm-line BS to the BSDOSPlotter, that's your fault.

Of course the points about auto-detecting symm line, etc. are valid, but I'm just chiming in on the band structure plotting which is the part I was involved with

**Comment by shyuep** [comment]
Note that I am not saying that I do not have the responsibility to pass the right object to the plotter (whether the new one or old one). My point is merely that *default parameters* do not result in a plot. Instead, I have to figure out there is a BS and a BSSymmLine (the differences are not entirely clear to me, when do we actually use the Non Symm Line version?), one of which is going to work, and to get that, I have to specify myriad options over multiple methods, and there are absolutely no examples.

All I am asking for is a "it just works" option for the majority of users. Right now, that is not the case.

**Comment by computron** [comment]
As far as I can tell, the changes you are suggesting are:

1. Vasprun.get_band_structure() should just ditch the line_mode parameter. Instead it should auto-detect whether we have a line mode and return a BandStructureSymmLine object if that's the case, or regular BandStructure object if not. Note that the latter is a subclass of the former so you get a BandStructure object no matter what.
2. The default symmetry in BandStructure should be 0.1(?)
3. There should be some examples of usage. e.g., starting from a VASP output dir and giving you a band structure plot

Is that correct?

**Comment by shyuep** [comment]
Yes for a start. Also, why is there a BS and a BSSymmLine. When is the former used?

**Comment by hautierg** [comment]
Anywhere you want to compute something from band structure (let say some electronic transport properties) you should use a band structure on an uniform band structure.
The band structures along symmetry lines are basically mainly for plotting.

**Comment by computron** [comment]
From a code (rather than science) perspective

- If you have k-points and energies, you have a BandStructure object. This is true both for symmetry line band structures as well as uniform mesh band structures.

**Comment by computron** [comment]
sorry I didn't finish the last comment, let me try again shortly...

**Comment by computron** [comment]
From a code (rather than science) perspective

* If you have k-points and energies, you (essentially) have a BandStructure object. This is true both for symmetry line band structures as well as uniform mesh band structures.
* If you can label certain k-points with names, and can define a sequential path through those k-points with distances, then you have special powers that are associated only with a line mode band structure.

There are multiple ways to structure that in code. The way that was chosen (not by me, although I don't think it's so bad) is to have BandStructure be the base class and BandStructureSymmLine be a subclass that inherits from it.

Of course, other code designs are possible. e.g., we could put everything in a single BandStructureClass and have certain variables like labels and distances be set to None if those didn't exist. And there can be an attribute called "is_line_mode" or something that automatically returns T/F depending on whether those variables were initialized or not. This would be useful for certain codes to be able to check what style of band structure was run (e.g., you often want to enforce either a uniform mesh or line mode mesh for a certain analysis). I am certainly OK doing it either way - e.g. class/subclass or all in one class.

**Comment by shyuep** [comment]
Ok. I see the code and science use cases now. For now, let's just simplify the usage of these by having some form of auto-detection and examples. I am fine with the class structure as it is for now.
